uniffi/src/ffi_converter_impls.rs

```diff
@@ -0,0 +1,409 @@
+};
+/// This module contains builtin `FFIConverter` implementations.  These cover:
+///   - Simple privitive types: u8, i32, String, Arc<T>, etc
+///   - Composite types: Vec<T>, Option<T>, etc.
+///   - SystemTime and Duration, which maybe shouldn`t be built-in, but have been historically and
+///     we want to continue to support them for now.
+///
+/// As described in
+/// https://mozilla.github.io/uniffi-rs/internals/lifting_and_lowering.html#code-generation-and-the-fficonverter-trait,
+/// we use the following system:
+///
+///   - Each UniFFIed crate defines a unit struct named `UniFFIConverter`
+///   - We define an `impl FFIConverter<T> for UniFFIConverter` for each `T` that we want to pass
+///     across the FFI.
+///   - When generating the code, we use `<crate::UniFFIConverter as ::uniffi::FFIConverter<T>>` to
+///     lift/lower/serialize types.
+///
+/// This crate needs to implement `FFIConverter<T>` on the `UniFFIConverter` instances for all
+/// UniFFIed crates, so it defines blanket impls like `impl FFIConverter<u8> for FC`.  "FC" means
+/// an abitrary FFIConverter type and these impls cover all the `UniFFIConverter` instances defined
+/// in the scaffolding code.
+use anyhow::bail;
+use bytes::buf::{Buf, BufMut};
+use paste::paste;
+use std::{
+    collections::HashMap,
+    convert::TryFrom,
+    time::{Duration, SystemTime},
+};
+
+/// Blanket implementation of `FfiConverter` for numeric primitives.
+///
+/// Numeric primitives have a straightforward mapping into C-compatible numeric types,
+/// sice they are themselves a C-compatible numeric type!
+macro_rules! impl_via_ffi_for_num_primitive {
+    ($($T:ty,)+) => { impl_via_ffi_for_num_primitive!($($T),+); };
+    ($($T:ty),*) => {
+            $(
+                paste! {
+                    unsafe impl<FC> FfiConverter<$T> for FC {
+                        type FfiType = $T;
+
+                        fn lower(obj: $T) -> Self::FfiType {
+                            obj
+                        }
+
+                        fn try_lift(v: Self::FfiType) -> Result<$T> {
+                            Ok(v)
+                        }
+
+                        fn write(obj: $T, buf: &mut Vec<u8>) {
+                            buf.[<put_ $T>](obj);
+                        }
+
+                        fn try_read(buf: &mut &[u8]) -> Result<$T> {
+                            check_remaining(buf, std::mem::size_of::<$T>())?;
+                            Ok(buf.[<get_ $T>]())
+                        }
+                    }
+                }
+            )*
+    };
+}
+
+impl_via_ffi_for_num_primitive! {
+    i8, u8, i16, u16, i32, u32, i64, u64, f32, f64
+}
+
+/// Support for passing boolean values via the FFI.
+///
+/// Booleans are passed as an `i8` in order to avoid problems with handling
+/// C-compatible boolean values on JVM-based languages.
+unsafe impl<FC> FfiConverter<bool> for FC {
```

Comment:
It's slightly weird that this is defining `FfiConverter<bool>` for all types, when we only need to target the `UniFFIConverter` unit structs defined in the scaffolding code.

To avoid that, we could define a trait named something like `FfiConverterDefault` and add `FC: FfiConverterDefault` bounds in these impls.  I wasn't sure if that simplified things or added confusion though.